MongoDB User Data Provider Support

MostafaMohamed2002 opened the pull request:
- Added UsersMongoImpl class for MongoDB storage
- Changed DataProvider interface to use suspend functions
- Modified repositories and use cases to work with coroutines
- Refactored all user tests to work with coroutines

MostafaMohamed2002 committed:
Add UsersMongoImpl with CRUD operations for user management

MostafaMohamed2002 committed:
add PlanMateException with MongoDB-specific exceptions

MostafaMohamed2002 committed:
Add MongoDB Exception Handler

MostafaMohamed2002 committed:
fix the id issue

Yaseen20022 commented on src/main/kotlin/utils/MongoExceptionHandler.kt:
```diff
@@ -0,0 +1,62 @@
+        throw when (e) {
+            is MongoWriteException -> {
+                when (e.error.code) {
+                    11000 -> PlanMateException.DuplicateKeyException(
```
make this number a const

Mostafa-alsaygh commented on src/main/kotlin/data/mongo/UsersMongoImpl.kt:
```diff
@@ -0,0 +1,93 @@
+import java.util.*
+
+class UsersMongoImpl(
+    private val db: MongoDBClient
```
db??
use good naming

Yaseen20022 commented on src/main/kotlin/utils/PlanMateException.kt:
```diff
@@ -22,4 +22,29 @@ open class PlanMateException(message: String) : Exception(message) {
     class InvalidStateIdException(message: String = "Invalid state id, no audit logs found.") :
         PlanMateException(message)
+
+    // MongoDB Exceptions
```
I suggest adding DatabaseUnavailableException  which is useful when there is a network error or a the server is down, basically it is used when the server is not available

MostafaMohamed2002 replied:
There is a Database TimeOut exception

MostafaMohamed2002 committed:
Refactor user use case tests for coroutines

- Update tests to use coroutines with runTest
- Replace Result wrapper with direct exceptions
- Use coEvery/coVerify for suspend functions

MostafaMohamed2002 committed:
fix MongoDB  implementations for suspend functions

MostafaMohamed2002 committed:
add MongoDB user data provider support

MostafaMohamed2002 committed:
merge develop into feature/users-mongodb

Yaseen20022 reviewed (approved):
Great work!

Yaseen20022 reviewed (approved):
Great work!
